### ISSUE TITLE
splk-dsm - dictinct count host issue inconsistency when setting up a dcount_host treshold

### ISSUE THREAD

**Issue by guilhemmarchand** [trackme-limited/trackme-report-issues#1]
User reported:

I recognized an unexpected red status for a data source in TrackMe v2.
The dc host value is defined as 22 and alerted. I cross checked the data with different time spans and saw these 22 hosts. But the alert says, that there are only 21 hosts found.
I think the reason is the underlying search of the tracker. It calculates the dcount_host per second and then the maximum of the result table is taken into account. So the rule expects all hosts to send data within the same second. Is this intended behaviour? I could image this could lead to a lot of false alarms.

```
| tstats max(_indextime) as data_last_ingest, min(_time) as data_first_time_seen, max(_time) as data_last_time_seen, count as data_eventcount, dc(host) as dcount_host where index=* sourcetype=* (index=* OR index=_*) (index=*) (splunk_server=*) (sourcetype!=stash sourcetype!=*too_small sourcetype!=modular_alerts:trackme* sourcetype!=trackme:*) [| inputlookup trackme_allowed_indexes-demotenant.csv |  fields+ index] _index_earliest="-4h" _index_latest="+4h" by _time, index, sourcetype span=1s
| eval data_last_ingestion_lag_seen=data_last_ingest-data_last_time_seen
``` intermediate calculation ```
| eval spantime=_time | bucket spantime span=5m | eventstats sum(data_eventcount) as eventcount_5m, avg(data_last_ingestion_lag_seen) as latency_5m by spantime, index, sourcetype
| stats latest(eventcount_5m) as latest_eventcount_5m, avg(eventcount_5m) as avg_eventcount_5m, stdev(eventcount_5m) as stdev_eventcount_5m, perc95(eventcount_5m) as perc95_eventcount_5m, latest(latency_5m) as latest_latency_5m, avg(latency_5m) as avg_latency_5m, stdev(latency_5m) as stdev_latency_5m, perc95(latency_5m) as perc95_latency_5m, max(data_last_ingest) as data_last_ingest, min(data_first_time_seen) as data_first_time_seen, max(data_last_time_seen) as data_last_time_seen, avg(data_last_ingestion_lag_seen) as data_last_ingestion_lag_seen, sum(data_eventcount) as data_eventcount, max(dcount_host) as dcount_host by index, sourcetype | eval data_last_ingestion_lag_seen=round(data_last_ingestion_lag_seen, 0) | rename index as data_index, sourcetype as data_sourcetype | eval object=data_index . ":" . data_sourcetype
``` tenant_id ```

```

**Comment by guilhemmarchand** [comment]
Under review

**Comment by sebwurl** [comment]
Is there already a change in the latest version? As I understand it, the KPI that is compared against the threshold is still biased by the tracker aggregation over 1s (v2.0.7)

**Comment by guilhemmarchand** [comment]
Hi @sebwurl !

Happy you ask ;-)

From Version 2.0.5 onward, first, you need to re-create the hybrid trackers to have the benefit from the new KPIs accordingly.

Once you have the fresh definiton, we have the following metrics:

<img width="1611" alt="image" src="https://user-images.githubusercontent.com/7503059/220479385-e4fc5fee-449f-49de-81a3-7117bf44ce91.png">

These metrics are calculated by the tracker as follows:

<img width="1770" alt="image" src="https://user-images.githubusercontent.com/7503059/220479671-67b12f6d-e589-4470-abcb-acee46a1f2e7.png">

So, these are:

<img width="1770" alt="image" src="https://user-images.githubusercontent.com/7503059/220479753-a52682e7-81b6-40db-9fc6-0039358b2a09.png">

The field "dcount_host" which is primarly used by the min dc count host is in fact the latest distinct count host per 5 minutes bucket of data for the current executon of the tracker.

As you can see in the screenshot above, the tracker first get the data on a per x plus host basis, then calculayes the dc host, then uses the latest for the KPI over the past 5 minutes and generates additional KPIs stored in the metric store.

In addition, you can leverage the ML to generate and maintain ML models against any of these KPIs, shall you want to do so.

In short, and thanks to you, we've addressed this effectively, the dc host is not based on the per sec but based on the latest dc host for the past 5 minutes, as the basis calculation.

Hope this makes sense - it sounded to us as a superior approach to do so

**Comment by sebwurl** [comment]
Re-creating the tracker makes totally sense, I could have tried that out... :)
Generally I can comprehend the new approach with the new grouping. Nevertheless I'm experiencing values like this in sourcegtype=trackme:state
![image](https://user-images.githubusercontent.com/85566586/220620911-aafcadeb-9612-4420-bb0c-e5e684ab2b54.png)
![image](https://user-images.githubusercontent.com/85566586/220620995-3e107a2b-4825-4935-af82-9c27c1db1334.png)

I don' think this is intended.
The search uses an avg value to calculate the latest_dcount_host_5m and dcount_host. I marked the flow in green:
![image](https://user-images.githubusercontent.com/85566586/220621855-66255cee-a2c4-47ca-b78a-107f0a59610b.png)

**Comment by guilhemmarchand** [comment]
@sebwurl 

The idea is to turn the number of dc hosts into a KPI, and leverage this KPI first the minimal dcount, and as well for ML purposes.

So, indeed, you cannot have a "float" dc host, technically it's an integrer, but as KPI, the calculation logic is:

- split the table by host
- from the table, get the dc(host) per index/sourcetype/<custom fields> (per min)

Then:
- use eventstats against a bucket time of 5 min to calculate the average dc(host) per entity
- use stats latest to take into account the latest calculation for the latest per 5 min bucket of data
- recycle this value as the currently known amount of dc hosts for that entity, to be stored in the KV for further verification purpses.

So, in fact, you ineed end up with a floating value, as it's a calculated value that will vary, then using the KPI metrics you can set this to a value that makes sense for your context.

Let me know if you are not convinced ;-)
